docs/question.md

```diff
@@ -204,14 +206,17 @@ def grade(data):
     # We can modify or delete any of these if we have a custom grading method.
     # This function only runs if `parse()` did not produce format errors, so we can assume all data is valid.
 
-    # grade() can also set `data['format_errors'][NAME]` if there is any reason to mark the question
+    # grade(data) can also set data['format_errors'][NAME] if there is any reason to mark the question
     # invalid during grading time.  This will cause the question to not use up one of the student's attempts' on exams.
-
-    # As an example, we will give half points for incorrect answers larger than "x":
-    if data["score"] == 0: # only if not already correct
-        if data["submitted_answers"]["y"] > data["params"]["x"]:
-            data["partial_scores"]["y"]["score"] = 0.5
-            data["score"] = 0.5
+    # You are encouraged, though, to do any checks for invalid data that can be done in `parse(data)` there instead,
+    # since that method is also called when the student hits "Save only" or in assessments without real-time grading.
+
+    # As an example, we will give half points for incorrect answers larger than "x",
+    # only if not already correct. Use math.isclose to avoid possible floating point errors.
+    if math.isclose(data["score"], 0.0) and data["submitted_answers"]["y"] > data["params"]["x"]:
+        data["partial_scores"]["y"]["score"] = 0.5
+        data["score"] = 0.5
+        data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
```

Comment:
With #6798 in mind this may need to be moved to partial scores. Same for the question.html code above.

Reply:
#6798 only applies to element code, not custom `server.py` files. I believe that the `feedback` entry in the data dictionary is specifically for use by custom question grader code, and the feedback in partial scores is for element grader code.

This is pretty confusing though, so I'll add some discussion about this somewhere.